test/unit/test_manifest_download.py

```diff
@@ -0,0 +1,82 @@
+import os
+import shutil
+import tempfile
+import unittest
+
+import scandir
+from mock import patch
+
+from hca.dss import DSSClient
+
+
+def _touch_file(path):
+    try:
+        os.makedirs(os.path.split(path)[0])
+    except OSError as e:
+        if 'File exists' in str(e):
+            pass
+    with open(path, 'w'):
+        pass
+
+
+def _fake_download(*args, **kwargs):
+    _touch_file(args[2])
+
+
+class TestManifestDownload(unittest.TestCase):
+    manifest = [
+        ['bundle_uuid', 'bundle_version', 'file_content_type', 'file_name', 'file_sha256', 'file_size', 'file_uuid', 'file_version', 'file_indexed'],
+        ['a_uuid', '1_version', 'somestuff', 'a_file_name', 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d', '12', 'af_uuid', 'af_version', 'False'],
+        ['b_uuid', '1_version', 'somestuff', 'b_file_name', '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4', '2', 'bf_uuid', 'af_version', 'False'],
+        ['c_uuid', '1_version', 'somestuff', 'c_file_name', '8f3404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c', '41', 'cf_uuid', 'af_version', 'False'],
+    ]
+
+    def setUp(self):
+        self.prev_wd = os.getcwd()
+        self.tmp_dir = tempfile.mkdtemp()
+        os.chdir(self.tmp_dir)
+        self.dss = DSSClient()
+        with open('manifest.tsv', 'w') as f:
+            f.write('\n'.join(['\t'.join(row) for row in self.manifest]))
+
+    def tearDown(self):
+        os.chdir(self.prev_wd)
+        shutil.rmtree(self.tmp_dir)
+
+    def _assert_all_files_downloaded(self):
+        num_files = sum(len(files) for _, _, files in scandir.walk('.'))
+        self.assertEqual(num_files, len(self.manifest))  # -1 for header, but +1 for manifest.tsv in cwd
+        expected_file_paths = [self.dss._file_path(row[4]) for row in self.manifest[1:]]
+        for path in expected_file_paths:
+            self.assertTrue(os.path.exists(path))
+
+    @patch('hca.dss.DSSClient.DIRECTORY_NAME_LENGTHS', [1, 2, 3])
+    def test_file_path(self):
+        self.assertRaises(AssertionError, self.dss._file_path, 'a')
+        parts = self.dss._file_path('abcdefghij').split(os.sep)
+        self.assertEqual(parts, ['a', 'bc', 'def', 'abcdefghij'])
+
+    @patch('hca.dss.DSSClient._download_file', side_effect=_fake_download)
+    def test_manifest_download(self, download_func):
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        # Since files now exist, running again ensures that we avoid unnecessary downloads
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        self._assert_all_files_downloaded()
+
+    @patch('hca.dss.DSSClient._download_file', side_effect=_fake_download)
+    def test_manifest_download_partial(self, _):
+        _touch_file(self.dss._file_path(self.manifest[1][4]))
```

Comment:
These magic column indices are annoying but eliminating them would require a more involved refactoring.

Reply:
I can refactor to use context manager if you want.

Reply:
Not sure how a CM would help here. Maybe using a `namedtuple` for the manifest rows would help. We could even use that namedtuple in the production code to avoid these magic  constants. You could use https://docs.python.org/2/library/collections.html#collections.somenamedtuple._fields to generate the header row.

You call, honestly.

Reply:
Your call if and how you want to do this.

Reply:
Sorry I was looking at the wrong line. Yes I would like to just leave this if you don't mind.